### DISCUSSION TITLE
Handling Cold Server Redundancy

### DISCUSSION THREAD

**Discussion by FSangouard** [eclipse-milo/milo#1227] [General]
Hello,

I am working on a client and I need to handle a cluster of servers where there is only one active server at any time, we can't connect to the backup server unless it becomes active during a failover (Cold Failover).

I was wondering what the right approach would be. I'm thinking of adding a SessionActivityListener to detect failure of the primary server and connect to the backup server. Here are my questions :

1. Do I need to have multiple instances of OpcUaClient, one for each server?
2. If yes, in the onSessionInactive() callback, do I need to call disconnect() on the client connected to the failed server and call connect() on the client configured to connect to the backup server?
3. If no, how can I switch servers while keeping the same instance of OpcUaClient?
4. Should I be prepared for transient failures which do not trigger a server failover?
5. If yes, how can I let the client try to reconnect automatically for a period of time before switching to the backup server? Can I do that in onSessionInactive()? Is it even a viable approach?

I'm currently still on v0.6.3, but if recent versions added features relevant to this question, I can upgrade.

Regards.

**Comment by kevinherron** [comment]
You will need multiple instances of `OpcUaClient`, one for each server.

> in the onSessionInactive() callback, do I need to call disconnect() on the client connected to the failed server and call connect() on the client configured to connect to the backup server?

Yes, in order to get an `OpcUaClient` instance to stop trying to reconnect and re-establish a Session, you'll have to disconnect it.

> Should I be prepared for transient failures which do not trigger a server failover?

I guess so? The client will automatically attempt to reconnect until `OpcUaClient::disconnect` is called.

> If yes, how can I let the client try to reconnect automatically for a period of time before switching to the backup server? Can I do that in onSessionInactive()? Is it even a viable approach?

I'm not sure anything in the library will help you with this. You'll have to track how long you've been disconnected for and decide when it's time to try another server in the set as part of your application logic. The spec leaves that up to the implementation to decide.

**Comment by FSangouard** [comment]
Thank you for your quick response!

I think the path forward is clear to me. 👍
